earth_enterprise/src/fusion/autoingest/sysman/khResourceProvider.cpp

```diff
@@ -645,12 +648,34 @@ khResourceProvider::JobLoop(StartJobMsg start)
       StartLogFile(job, start.logfile);
     }
 
-    success = RunCmd(job, jobid, start.commands[cmdnum], cmdnum == 0, cmdtime, endtime, logTotalTime);
-    if (!Valid(job)) return;  // check if somebody already asked for me to go away
+    success = false;
+    for (uint tries = 0; tries < cmdTries && !success; ++tries) {
+      if (tries > 0) {
+        // Write out the overall time if we run a command more than once.
+        logTotalTime = true;
+        if (job->logfile) {
+          LogRetry(job, tries, cmdTries, sleepBetweenTriesSec);
+        }
+        if (sleepBetweenTriesSec > 0) {
+          {
+            // Release the lock while we sleep
+            khUnlockGuard unlock(mutex);
+            sleep(sleepBetweenTriesSec);
+          }
+          // Once we have the lock again, check if someone deleted the job while
+          // we were sleeping
+          job = FindJobById(jobid);
+          if (!Valid(job)) return;
+        }
+      }
+      success = RunCmd(job, jobid, start.commands[cmdnum], cmdtime, endtime, progressSent);
+      if (!Valid(job)) return;  // check if somebody already asked for me to go away
```

Comment:
What would cause `job` to be invalid right after `RunCmd`? Being canceled by a user?

Reply:
I think a user canceling it would be an example of how this could happen. Based on the old code, if we release the lock we first have to check if the job is still valid after re-acquiring the lock. This particular instance is actually a continuation of the check on line 729 - we release the lock to wait for the command to run, then on line 729 we check to see if the job is still valid after getting the lock again. If it's not, we return to this point in the code, so we check again if it's valid to distinguish between cases where the job completed unsuccessfully and the job was canceled.